### PR TITLE
Make code fix to add 'this.' work for statics

### LINKED ISSUE #23519: Quick Fix for adding `this.` to undefined static property referenced in static method
From https://github.com/Microsoft/vscode/issues/48104

**TypeScript Version:**  2.9.0-dev.20180418

**Search Terms:** 
- quickfix
- code action
- static
- this

**Code**

```ts
class foo {
    static bar() { }

    static test() {
        bar()
    }
}
```

**Suggestion:**
Hover over the error in `bar()`.  See this suggestion in the error message:
screen shot 2018-04-18 at 11 42 54 am

But no quick fix is available:

https://github.com/Microsoft/vscode/issues/48104 suggestions that the error message is also wrong and we should use `this.bar` instead of `foo.bar`

**Playground Link:**  

**Related Issues:**

### LINKED ISSUE #23520: Use 'this.' insertText for static properties completions
From https://github.com/Microsoft/vscode/issues/48104

**TypeScript Version:**  2.9.0-dev.20180418

**Search Terms:** 
- quickfix
- code action
- static
- this

**Code**

```ts
class foo {
    static bar() { }

    static test() {
        
    }
}
```

1. Start typing `bar` in `test`

**Current behavior:**
`bar` is shown as a suggestion. Accepting it completes to `bar()` which produces an error

**Suggested behavior:**
Use the `insertText` completionItem property to make the entry for `bar` complete to `this.bar`, just like it would if these methods were not static

**Playground Link:**  

**Related Issues:**

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #23519

### CODE REVIEW

**Comment by amcasey** [comment]
I would have guessed we'd prefer to use the name of the declaring type as the receiver (`C.m` in your example).

**Comment by andy-ms** [comment]
If we do that, we should probably change the completions to work the same way. Although the error message does use the class name instead of `this`, possibly because it also applies inside instance methods. (So, another issue is to make the code fix add the class name inside an instance method.) @DanielRosenwasser this was last touched in #6493, thoughts?

**Comment by DanielRosenwasser** [comment]
I picked the class name because

1. It works within both static and instance contexts.
2. `this` is never guaranteed to have the same shape between any two classes, which makes it feel odd.

File: src/services/codefixes/fixForgottenThisPropertyAccess.ts [hunk 1]
```diff
@@ -1,35 +1,36 @@
...
+            doChange(changes, context.sourceFile, getInfo(diag.file, diag.start!, diag.code));
         }),
     });
 
-    function getNode(sourceFile: SourceFile, pos: number): Identifier | undefined {
+    interface Info { readonly node: Identifier; readonly className: string | undefined; }
+    function getInfo(sourceFile: SourceFile, pos: number, diagCode: number): Info | undefined {
         const node = getTokenAtPosition(sourceFile, pos, /*includeJsDocComment*/ false);
-        return isIdentifier(node) ? node : undefined;
+        return isIdentifier(node) ? { node, className: diagCode === errorCodes[1] ? getContainingClass(node).name.text : undefined } : undefined;
```
Comment (DanielRosenwasser) [inline src/services/codefixes/fixForgottenThisPropertyAccess.ts]: This breaks if someone ever adds to the front of that list. Just provide a `const` at the top of the file like

```ts
const didYouMeanStaticMemberCode = Diagnostics.Cannot_find_name_0_Did_you_mean_the_static_member_1_0.code,
```

Comment (DanielRosenwasser) [reply src/services/codefixes/fixForgottenThisPropertyAccess.ts]: Also, as a nit, I'd expand these out to statements. It gets a little hard to unfold ever branch within this single line.

**Comment by andy-ms** [comment]
Made an issue #23539 for aligning this with completions

**Comment by andy-ms** [comment]
@DanielRosenwasser Could you update your review?

**Pull request merged**